resource-aggregate/cqrs/eventstore/cqldb/load.go

```diff
@@ -0,0 +1,228 @@
+	q.WriteString("select * from " + s.Table())
+	if filter != "" {
+		q.WriteString(" " + cqldb.WhereClause + " " + filter)
+	}
+	q.WriteString(";")
+	iter := s.client.Session().Query(q.String()).WithContext(ctx).Iter()
+	i := newIterator(iter, s.config.unmarshalerFunc)
+	err := eh.Handle(ctx, i)
+	errClose := iter.Close()
+	if err == nil {
+		return errClose
+	}
+	return err
+}
+
+// LoadUpToVersion loads aggregates events up to a specific version. It loads events from the last snapshot. Only last snapshots
+func (s *EventStore) LoadUpToVersion(_ context.Context, _ []eventstore.VersionQuery, _ eventstore.Handler) error {
+	return eventstore.ErrNotSupported
+}
+
+// LoadFromVersion loads aggregates events from version.
+func (s *EventStore) LoadFromVersion(ctx context.Context, queries []eventstore.VersionQuery, eh eventstore.Handler) error {
+	q := make([]eventstore.SnapshotQuery, 0, len(queries))
+	for _, query := range queries {
+		q = append(q, eventstore.SnapshotQuery{
+			GroupID:     query.GroupID,
+			AggregateID: query.AggregateID,
+		})
+	}
+	return s.LoadFromSnapshot(ctx, q, eh)
+}
+
+func snapshotQueriesToFilter(deviceID string, queries []eventstore.SnapshotQuery, timestamp int64) string {
+	var filter strings.Builder
+	if deviceID != "" {
+		filter.WriteString(deviceIDKey + "=" + deviceID)
+	}
+	aggrs := make([]string, 0, len(queries))
+	for _, q := range queries {
+		if q.AggregateID != "" {
+			aggrs = append(aggrs, q.AggregateID)
+		}
+	}
+	if len(aggrs) > 0 {
+		if filter.Len() > 0 {
+			filter.WriteString(" and ")
+		}
+		filter.WriteString(idKey)
+		filter.WriteString(" in (")
+		for idx, aggr := range aggrs {
+			if idx > 0 {
+				filter.WriteString(",")
+			}
+			filter.WriteString(aggr)
+		}
+		filter.WriteString(")")
+	}
+	if timestamp > 0 {
+		if filter.Len() != 0 {
+			filter.WriteString(" and ")
+		}
+		filter.WriteString(timestampKey)
+		filter.WriteString(">=")
+		filter.WriteString(fmt.Sprintf("%v", timestamp))
+		filter.WriteString(" ALLOW FILTERING")
+	}
+	return filter.String()
+}
+
+func (s *EventStore) loadFromSnapshotByGroup(ctx context.Context, groupID string, queries []eventstore.SnapshotQuery, timestamp int64, eventHandler eventstore.Handler) error {
+	if groupID != "" {
+		if _, err := uuid.Parse(groupID); err != nil {
+			return fmt.Errorf("invalid groupID %v: %w", groupID, err)
+		}
+	}
+	for _, query := range queries {
+		if _, err := uuid.Parse(query.AggregateID); err != nil {
+			return fmt.Errorf("invalid aggregateID %v: %w", query.AggregateID, err)
+		}
+	}
+
+	filter := snapshotQueriesToFilter(groupID, queries, timestamp)
+	return s.loadEventsQuery(ctx, eventHandler, filter)
+}
+
+func normalizeQueries(queries []eventstore.SnapshotQuery) map[string][]eventstore.SnapshotQuery {
+	normalizeQuery := make(map[string][]eventstore.SnapshotQuery, len(queries))
+	for _, query := range queries {
+		if query.AggregateID == "" {
+			normalizeQuery[query.GroupID] = make([]eventstore.SnapshotQuery, 0, 1)
```

Comment:
make([]eventstore.SnapshotQuery) doesn't seem like the 1 capacity is needed

Reply:
fixed